### PR TITLE
Add the Slider component

### PR DESCRIPTION
**Pull request by ynotdraw**
## 🚀 Description

Adds a new Slider component.

For the most part, it mimics how `<input type="range"` behaves, except:

- There is a `multiple` mode.
- Rather than defaulting to 50% of the range size, single mode does 25% and multiple mode does a 25/75% split.

## 📋 Checklist

- I have followed the [Contributing Guidelines](https://github.com/crowdstrike/glide-core/blob/main/CONTRIBUTING.md).
- I have added tests to cover new or updated functionality.
- I have added or updated Storybook stories.
- I have [localized](https://github.com/CrowdStrike/glide-core/blob/main/CONTRIBUTING.md#translations-and-static-strings) new strings.
- I have followed the [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/) or met with the Accessibility Team.
- I have included a [changeset](https://github.com/CrowdStrike/glide-core/blob/main/CONTRIBUTING.md#versioning-a-package).
- I have scheduled a design review.
- I have reviewed the Storybook and Visual Test Report links below.

## 🔬 Testing

Tests should have us covered, but for some manual validations:

**Single mode**

- Go to [Storybook](https://glide-core.crowdstrike-ux.workers.dev/slider?path=/story/slider--slider).
- Press and hold the mouse on the handle, begin dragging.
- Verify you can drag the handle and the value in the input updates.

- Go to [Storybook](https://glide-core.crowdstrike-ux.workers.dev/slider?path=/story/slider--slider).
- Type a value into the input.
- Verify the drag handle position updates.

- Go to [Storybook](https://glide-core.crowdstrike-ux.workers.dev/slider?path=/story/slider--slider).
- Turn on Voice Over.
- Tab to the handle.
- Verify the `label` is read aloud.
- Press `control` + `option` + shift` + `arrowdown`.
- Verify Voice Over reads "in slider".
- Press `arrowright`.
- Verify the new value is announced.
- Tab to the input field now.
- Verify the current value and `label` is announced.

**Multiple mode**

- Go to [Storybook](https://glide-core.crowdstrike-ux.workers.dev/slider?path=/story/slider--slider&args=multiple:!true)
- Press and hold the mouse on the "minimum handle" (the one on the left), begin dragging.
- Verify you can drag the handle and the value in the minimum input (the one on the left) updates.
- Now do the same thing, but for the "maximum handle" (the one on the right).

- Go to [Storybook](https://glide-core.crowdstrike-ux.workers.dev/slider?path=/story/slider--slider&args=multiple:!true)
- Type a value into the minimum input.
- Verify the minimum drag handle position updates.
- Do the same thing for the maximum input.

- Go to [Storybook](https://glide-core.crowdstrike-ux.workers.dev/slider?path=/story/slider--slider&args=multiple:!true)
- Turn on Voice Over.
- Tab to the minimum input.
- Verify the minimum input value and "Set minimum `label`" is announced.
- Tab to the minimum handle.
- Verify the "Minimum `label`" is read aloud.
- Press `control` + `option` + shift` + `arrowdown`.
- Verify Voice Over reads "in slider".
- Press `arrowright`.
- Verify the new value is announced.
- Do the same thing for the maximum handle and input, except it should read aloud "maximum".

### CODE REVIEW

File: src/input.test.forms.ts
Comment (ynotdraw) [inline src/input.test.forms.ts:1]: Some naming oddities I cleaned up and aligned with Slider. Let me know if we shouldn't do this or if you have better suggestions.

File: src/slider.styles.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  css`
+    ${focusOutline('.handle:focus-visible')}
+  `,
+  css`
+    /* The designs call for a bit more spacing than the default when vertical. */
+    glide-core-private-label[orientation='vertical']::part(private-tooltips) {
+      margin-block-end: var(--glide-core-spacing-base-xxs);
+    }
+
+    .slider-container {
```
Comment (ynotdraw) [inline src/slider.styles.ts:14]: I wanted to name this simply `component`, but it's not at the root level. So that doesn't quite make sense given our other patterns. Open to suggestions!

File: src/slider.styles.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+      &.single {
+        /*
+            Without this suggestion from design, when the handle
+            is at the min, it would overflow the container and
+            not align with label. This isn't needed in multiple
+            mode because we have an input to help prevent this
+            from happening.
+         */
+        padding-inline-start: var(--glide-core-spacing-base-sm);
```
Comment (ynotdraw) [inline src/slider.styles.ts:84]: Hopefully this makes sense.

Without the additional padding 👎 

Screenshot 2025-05-20 at 10 28 41 AM

Screenshot 2025-05-20 at 10 29 03 AM

File: src/slider.ts [hunk 3]
```diff
@@ -0,0 +1,1331 @@
...
+  // Intentionally not reflected to match native.
+  /**
+   * @default []
+   */
+  @property({ type: Array })
+  get value(): number[] {
+    if (
+      this.multiple &&
+      this.minimumValue !== undefined &&
+      this.maximumValue !== undefined
```
Comment (ynotdraw) [inline src/slider.ts:96]: `!== undefined` rather than a truthy check, because we need to allow for `0`

File: src/slider.ts [hunk 4]
```diff
@@ -0,0 +1,1331 @@
...
+    ) {
+      return [this.minimumValue, this.maximumValue];
```
Comment (ynotdraw) [inline src/slider.ts:98]: This is a different pattern, but rather than having `#value` and keeping it as an array, I found it easier to maintain, update, and read by having two distinctly separate variables for our values. Especially because the single mode only has one.

File: src/slider.styles.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      inline-size: 1.5rem;
+      inset-block-start: 50%;
+      position: absolute;
+      transform: translate(-50%, -50%);
+      transition:
+        border 150ms ease-in-out,
+        box-shadow 150ms ease-in-out;
+
+      &:not(.disabled, .readonly):active,
+      &:not(.disabled, .readonly):hover {
```
Comment (clintcs) [inline src/slider.styles.ts:138]:
```suggestion
      &:not(.disabled, .readonly):is(:active, :hover) {
```

File: src/input.test.forms.ts [hunk 6]
```diff
@@ -36,7 +36,7 @@ it('can be reset if there was no initial value', async () => {
   expect(host.value).to.be.empty.string;
 });
 
-it('has `formData` value when it has a value', async () => {
+it('has a `formData` value when it has a value', async () => {
```
Comment (clintcs) [inline src/input.test.forms.ts:39]: Is a `formData` a thing that a component or element has?

File: src/slider.stories.ts [hunk 7]
```diff
@@ -0,0 +1,283 @@
+import './icons/storybook.js';
```
Comment (clintcs) [inline src/slider.stories.ts:1]: Used?

File: src/slider.stories.ts [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    step: 1,
+    tooltip: '',
+    value: [],
+    version: '',
+  },
+  play(context) {
+    const slider = context.canvasElement.querySelector('glide-core-slider');
+
+    if (context.name.includes('Error') && slider instanceof SliderComponent) {
+      // Slider is a bit unique in that it always
```
Comment (clintcs) [inline src/slider.stories.ts:60]: Nit. Mind making the width of the this comment similar to your other comments?

File: src/slider.stories.ts [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+      // has a `value` set. To force an error state,
+      // we have to use `setCustomValidity`.
+      slider.setCustomValidity('Error');
```
Comment (clintcs) [inline src/slider.stories.ts:63]:
```suggestion
      slider.setCustomValidity('Invalid');
```

🤔

File: src/slider.stories.ts [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    name: '',
+    orientation: 'horizontal',
+    readonly: false,
+    'reportValidity()': '',
+    required: false,
+    'resetValidityFeedback()': '',
+    'setCustomValidity(message)': '',
+    'setValidity(flags, message)': '',
+    'slot="description"': '',
+    step: 1,
```
Comment (clintcs) [inline src/slider.stories.ts:51]: `step` is getting rendered in the code example. `orientation` is too.

File: src/slider.stories.ts [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    },
+    max: {
+      table: {
+        defaultValue: { summary: '"100"' },
+        type: { summary: 'string' },
+      },
+    },
+    min: {
+      table: {
+        defaultValue: { summary: '"0"' },
```
Comment (clintcs) [inline src/slider.stories.ts:160]:
```suggestion
        defaultValue: { summary: '0' },
```

File: src/slider.stories.ts [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+    },
+    'hide-label': {
+      table: {
+        defaultValue: { summary: 'false' },
+        type: { summary: 'boolean' },
+      },
+    },
+    max: {
+      table: {
+        defaultValue: { summary: '"100"' },
```
Comment (clintcs) [inline src/slider.stories.ts:154]:
```suggestion
        defaultValue: { summary: '100' },
```

File: src/slider.styles.ts [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+      */
+      &.readonly {
+        background-color: transparent;
+        border: 1px solid transparent;
+        padding-inline-start: 0;
+      }
+    }
+
+    .track-container {
+      flex: 1;
```
Comment (clintcs) [inline src/slider.styles.ts:73]:
```suggestion
      flex-grow: 1;
```

File: src/slider.styles.ts (hunk 2 above)
Comment (clintcs) [reply src/slider.styles.ts:84]: They're okay with the error message being misaligned as a result?

image

File: src/slider.styles.ts [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+      border-radius: var(--glide-core-rounding-base-radius-round);
+      inline-size: 100%;
+      position: relative;
+
+      &.disabled {
+        background-color: var(--glide-core-color-static-stroke-secondary);
+      }
+    }
+
+    .handle {
```
Comment (clintcs) [inline src/slider.styles.ts:120]: Didn't like `.thumb`?

File: src/slider.styles.ts (hunk 5 above)
Comment (clintcs) [inline src/slider.styles.ts:138]:
```suggestion
      &:not(.disabled, .readonly):is(:active, :hover) {
```

File: src/slider.test.aria.ts [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+  await expect(page.locator('glide-core-slider')).toMatchAriaSnapshot(`
+    - text: Label
+    - group:
+      - slider "Label"
+    - spinbutton "Label"
+  `);
+});
+
+test('multiple', async ({ page }) => {
```
Comment (clintcs) [inline src/slider.test.aria.ts:19]: Should this be somewhere in the `describe` block below?

File: src/slider.test.basics.multiple.ts [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+  );
+
+  const maximumInput = host.shadowRoot?.querySelector<HTMLInputElement>(
+    '[data-test="maximum-input"]',
+  );
+
+  expect(maximumInput?.getAttribute('max')).to.equal('200');
+});
+
+it('sets `max` on the minimum input based on the current maximum value - `step`', async () => {
```
Comment (clintcs) [inline src/slider.test.basics.multiple.ts:143]: Nit. `-` could be confused a dash. Might be better to spell it out as "minus".

File: src/slider.test.basics.single.ts [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+
+it('is accessible ', async () => {
+  const host = await fixture<Slider>(
+    html`<glide-core-slider label="Label"></glide-core-slider>`,
+  );
+
+  await expect(host).to.be.accessible();
+});
+
+it('sets `value` when one is not provided to 25%', async () => {
```
Comment (clintcs) [inline src/slider.test.basics.single.ts:12]: As is, it can be read two different ways.

```suggestion
it('sets `value` to 25% when not provided', async () => {
```

File: src/slider.test.events.multiple.ts [hunk 18]
```diff
@@ -0,0 +1,896 @@
+import { assert, aTimeout, expect, fixture, html } from '@open-wc/testing';
+import sinon from 'sinon';
+import { sendKeys } from '@web/test-runner-commands';
+import Slider from './slider.js';
+
+it('dispatches an "input" event typing in the minimum input', async () => {
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:6]: Strictly speaking, _it_ (the component) doesn't type into the minimum input.

```suggestion
it('dispatches an "input" event when its minimum is changed', async () => {
```

File: src/slider.test.events.multiple.ts [hunk 19]
```diff
@@ -0,0 +1,896 @@
...
+  );
+
+  const spy = sinon.spy();
+  host.addEventListener('change', spy);
+
+  const minimumInput = host.shadowRoot?.querySelector<HTMLInputElement>(
+    '[data-test="minimum-input"]',
+  );
+
+  minimumInput?.focus();
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:56]: Tab to and away from instead of programmatically blurring? There are a few instances of this.

File: src/slider.test.events.multiple.ts [hunk 20]
```diff
@@ -0,0 +1,896 @@
...
+  );
+
+  maximumInput?.focus();
+
+  await sendKeys({ type: '1' });
+
+  expect(spy.callCount).to.equal(1);
+});
+
+it('dispatches a "change" event blurring the minimum input', async () => {
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:44]: Does the "change" event actually blur the minimum input? Or is "change" dispatched when the input is blurred?

File: src/slider.test.events.multiple.ts [hunk 21]
```diff
@@ -0,0 +1,896 @@
...
+  await sendKeys({ type: '1' });
+
+  expect(spy.callCount).to.equal(0);
+
+  maximumInput?.blur();
+
+  expect(spy.callCount).to.equal(1);
+});
+
+it('dispatches an "input" event dragging the minimum handle', async () => {
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:90]:
```suggestion
it('dispatches an "input" event when dragging the minimum handle', async () => {
```

File: src/slider.test.events.multiple.ts [hunk 22]
```diff
@@ -0,0 +1,896 @@
...
+  const minimumHandle = host.shadowRoot?.querySelector(
+    '[data-test="minimum-handle"]',
+  );
+
+  const trackRect = sliderTrack?.getBoundingClientRect();
+
+  assert(trackRect);
+  assert(minimumHandle);
+
+  minimumHandle.dispatchEvent(
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:109]: We should probably add "down" and "up" to `library/mouse.ts` and use that her.e

File: src/slider.test.events.multiple.ts [hunk 23]
```diff
@@ -0,0 +1,896 @@
...
+    new MouseEvent('mousedown', {
+      bubbles: true,
+      cancelable: true,
+    }),
+  );
+
+  await aTimeout(0);
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:116]: Wanna say why here and below?

File: src/slider.test.events.multiple.ts (hunk 21 above)
Comment (clintcs) [reply src/slider.test.events.multiple.ts:90]: There are other instances of "when" missing throughout.

File: src/slider.test.events.multiple.ts [hunk 24]
```diff
@@ -0,0 +1,896 @@
...
+  host.addEventListener('change', changeSpy);
+
+  const sliderTrack = host.shadowRoot?.querySelector('[data-test="slider"]');
+  const trackRect = sliderTrack?.getBoundingClientRect();
+
+  assert(trackRect);
+  assert(sliderTrack);
+
+  // Click on the track at the 90% position, which is closer
+  // to the maximum handle than the minimum handle.
```
Comment (clintcs) [inline src/slider.test.events.multiple.ts:359]: The "click" part is explicit in the code:

```suggestion
  // The 90% position is closer to the maximum handle than the minimum handle.
```

File: src/slider.test.visuals.ts [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+
+              element.append(div);
+            });
+
+          await expect(page).toHaveScreenshot(
+            `${test.titlePath.join('.')}.png`,
+          );
+        });
+
+        test(`value="['80']"`, async ({ page }, test) => {
```
Comment (clintcs) [inline src/slider.test.visuals.ts:201]: Would this test and the one below be better in your `multiple` block above or no?

File: src/slider.ts (hunk 3 above)
Comment (clintcs) [reply src/slider.ts:96]: I was just working on lint stuff. How do you feel about [this](https://typescript-eslint.io/rules/strict-boolean-expressions) rule?

File: src/slider.ts [hunk 26]
```diff
@@ -0,0 +1,1331 @@
...
+              @type {Element | string}
+            -->
+          </slot>
+
+          ${when(
+            this.#isShowValidationFeedback && this.validityMessage,
+            () =>
+              html`<span class="validity-message" data-test="validity-message"
+                >${unsafeHTML(this.validityMessage)}</span
+              >`,
```
Comment (clintcs) [inline src/slider.ts:675]:
```suggestion
            () => html`
              <span class="validity-message" data-test="validity-message">
                ${unsafeHTML(this.validityMessage)}
              </span>
            `,
```

File: src/slider.ts [hunk 27]
```diff
@@ -0,0 +1,1331 @@
...
+      // Track clicks can inadvertently occur when letting go of
+      // a handle. To help prevent click events from dispatching,
+      // this variable keeps track of when we are still processing
+      // the drag event.
+      //
+      // Resetting the state in the next frame allows the drag
+      // to fully complete before accepting track click events
+      // again.
+      setTimeout(() => {
+        this.#isCompletingDrag = false;
```
Comment (clintcs) [inline src/slider.ts:858]: "Completing" or "Completed"?

File: src/slider.ts [hunk 28]
```diff
@@ -0,0 +1,1331 @@
...
+          this.form?.requestSubmit();
+          return;
+        }
+
+        default: {
+          return;
+        }
+      }
+
+      event.preventDefault();
```
Comment (clintcs) [inline src/slider.ts:1078]: Is it obvious why the event is canceled?

File: src/slider.styles.ts (hunk 2 above)
Comment (ynotdraw) [reply src/slider.styles.ts:84]: Yeah. It looks off with a single slider for sure, but when `multiple` it's not bad because we have the input on the left-hand side too.

File: src/slider.ts (hunk 3 above)
Comment (ynotdraw) [reply src/slider.ts:96]: I think I dig it. Would help to prevent foot guns, I think. How about you?

Comment (clintcs) [reply src/slider.ts:96]: Same, yeah. Plus it's explicit. It's just usually a hard sell to developers, who often aren't used to it.

File: src/slider.styles.ts (hunk 14 above)
Comment (ynotdraw) [reply src/slider.styles.ts:120]: I actually had it as `.thumb` since the beginning, as that's what the platform calls it. But design refers to it as a "handle", so I opted for that instead.  I can ask if they'd be open to calling it a thumb instead!

File: src/slider.test.aria.ts (hunk 15 above)
Comment (ynotdraw) [reply src/slider.test.aria.ts:19]: A bit of another tricky aria naming situation.

I want to test when `multiple` is set, but nothing else.  The existing describe blocks below test with `multiple` and `disabled` or `readonly`.  So the naming is a bit... odd.

Maybe:

```js
test.describe('multiple=${true}', () => {
  test('disabled=${false}', async ({ page }) => {
```

Or even more explicit:

```js
test.describe('multiple=${true}', () => {
  test('disabled=${false}, readonly=${false}', async ({ page }) => {
```

What do you prefer/think?

File: src/slider.test.events.multiple.ts (hunk 19 above)
Comment (ynotdraw) [reply src/slider.test.events.multiple.ts]: I dig it - updating! I found myself reaching for `focus()` and `blur()` more for convenience, but tabbing to and away is more common in our code base. Thanks for the eyes!

File: src/slider.test.events.multiple.ts (hunk 20 above)
Comment (ynotdraw) [reply src/slider.test.events.multiple.ts]: > "change" dispatched when the input is blurred

updating! Thank you!

File: src/slider.test.events.multiple.ts (hunk 23 above)
Comment (ynotdraw) [reply src/slider.test.events.multiple.ts]: Thanks for pointing this out - turns out in most cases we only care about `await host.updateComplete`. There are only two edge case tests ( `prevents track click events dispatching immediately after completing a drag operation` and `prevents track click events dispatching immediately after completing a drag operation` ) where we do need `await aTimeout(0)` and those two areas have comments. Thank you!

File: src/slider.test.visuals.ts (hunk 25 above)
Comment (ynotdraw) [reply src/slider.test.visuals.ts]: I mostly left them here to align with [the Dropdown visual tests](https://github.com/CrowdStrike/glide-core/blob/main/src/dropdown.test.visuals.ts#L275-L304), but I do think it makes sense to put these in the appropriate blocks. Want me to update Dropdown too while I'm here?

File: src/slider.ts (hunk 28 above)
Comment (ynotdraw) [reply src/slider.ts:1083]: Comment added (thank you!).

File: src/slider.test.events.multiple.ts (hunk 22 above)
Comment (ynotdraw) [reply src/slider.test.events.multiple.ts:102]: Checking this out now!

**Comment by ynotdraw** [comment]
@clintcs thank you for the early review!

Because GitHub comments can be hard to track down, I believe I covered all of your comments.  But here are a few I'm curious your takes on (hopefully these links work and don't take you to an odd area of the PR 🤞 ):

- https://github.com/CrowdStrike/glide-core/pull/897#discussion_r2099028649
- https://github.com/CrowdStrike/glide-core/pull/897#discussion_r2099047544

File: src/slider.test.visuals.ts (hunk 25 above)
Comment (clintcs) [reply src/slider.test.visuals.ts]: I see. I leave it to you for Slider. But Dropdown is a bit different because it doesn't have a `describe` block specifically for `multiple`

File: src/slider.ts (hunk 27 above)
Comment (ynotdraw) [reply src/slider.ts]: Updated!

File: src/input.test.forms.ts (hunk 6 above)
Comment (clintcs) [reply src/input.test.forms.ts]: Sorry for the churn. I think your edit was right. For whatever reason, I skipped over the "value" part when reading it.

File: src/slider.stories.ts (hunk 8 above)
Comment (ynotdraw) [reply src/slider.stories.ts]: Should be taken care of now. I added some rulers in my editor to help

File: src/slider.stories.ts (hunk 10 above)
Comment (clintcs) [reply src/slider.stories.ts:50]: Ah. That's right. I [removed](https://github.com/CrowdStrike/glide-core/pull/850) the default argument stuff from `preview.js` because it was getting out of hand. We now handle in stories not rendering attributes at their defaults.

I'll push a commit.

Comment (clintcs) [reply src/slider.stories.ts:50]: https://github.com/CrowdStrike/glide-core/pull/897/commits/a8c310fa77dbe47b8d53e4a1e8908ce47820687d ✅

**Review by clintcs — APPROVED** [review]

File: src/slider.stories.ts [hunk 29]
```diff
@@ -0,0 +1,283 @@
...
+      },
+    },
+    'hide-label': {
+      table: {
+        defaultValue: { summary: 'false' },
+        type: { summary: 'boolean' },
+      },
+    },
+    max: {
+      table: {
```
[the remark this answers is not in the document]
Comment (ynotdraw) [reply src/slider.stories.ts:154]: ✅  https://github.com/CrowdStrike/glide-core/pull/897/commits/eba0597365c1afddac9dc838fe76196845e98f49

File: src/slider.stories.ts [hunk 30]
```diff
@@ -0,0 +1,283 @@
...
+        defaultValue: { summary: '100' },
+        type: { summary: 'string' },
```
[the remark this answers is not in the document]
Comment (ynotdraw) [reply src/slider.stories.ts]: ✅  https://github.com/CrowdStrike/glide-core/pull/897/commits/eba0597365c1afddac9dc838fe76196845e98f49

File: src/slider.stories.ts [hunk 31]
```diff
@@ -0,0 +1,283 @@
...
+      },
+    },
+    'slot="description"': {
+      table: {
+        type: { summary: 'Element' },
+      },
+    },
+    step: {
+      table: {
+        defaultValue: { summary: '"1"' },
```
[the remark this answers is not in the document]
Comment (ynotdraw) [reply src/slider.stories.ts]: ✅  https://github.com/CrowdStrike/glide-core/pull/897/commits/eba0597365c1afddac9dc838fe76196845e98f49

File: src/slider.test.visuals.ts (hunk 25 above)
Comment (ynotdraw) [reply src/slider.test.visuals.ts]: ✅ Went ahead and moved them with https://github.com/CrowdStrike/glide-core/pull/897/commits/0e357cd77a2905ee415c023d6bfbb097241e9496#diff-afcce9a479f24f6bc29ac9fe101d4c86ddc730e0ae2eddb5641fe5c99dec74ae

File: src/slider.stories.ts [hunk 32]
```diff
@@ -0,0 +1,286 @@
...
+  },
+  args: {
+    label: 'Label',
+    'addEventListener(event, handler)': '',
+    'checkValidity()': '',
+    disabled: false,
+    'hide-label': false,
+    max: '',
+    min: '',
+    multiple: false,
```
Comment (danwenzel) [inline src/slider.stories.ts:40]: Nit: curious about the choice to use the world `multiple` vs something like `range`.

File: src/slider.ts [hunk 33]
```diff
@@ -0,0 +1,1350 @@
...
+    }
+
+    if (this.minimumValue !== undefined) {
+      return [this.minimumValue];
+    }
+
+    return [];
+  }
+
+  set value(value: number[]) {
```
Comment (danwenzel) [inline src/slider.ts:108]: If I set the `value` to something silly like `[75, 25]`, I get an interesting result:

image

Maybe an edge case, and consumers should be in charge of validation? Not sure, but I thought I'd point it out

**Review by danwenzel — APPROVED** [review]

File: src/slider.stories.ts (hunk 32 above)
Comment (ynotdraw) [reply src/slider.stories.ts:40]: I chose `multiple` to align with Dropdown, as we tend to use the same attribute names for similar functionality. But I'm curious what you and @clintcs think too. Does `range` make more sense here?

Comment (clintcs) [reply src/slider.stories.ts:40]: I don't think it would be a problem to deviate from Dropdown. Dropdown uses`multiple` attribute because native uses `multiple` with `<select>`. And native ranges don't have a `multiple` attribute.

But I think `multiple` makes sense here. `range`, in terms of native, isn't an attribute in itself and, even with `type="range"`, it doesn't mean multiple ranges. It just means "slider".

File: src/slider.ts [hunk 34]
```diff
@@ -0,0 +1,1350 @@
...
+      const normalizedMaximum = Math.round(value[1] / this.step) * this.step;
+
+      // Clamp the normalized values to the allowed ranges.
+      this.minimumValue = Math.max(normalizedMinimum, this.min);
+      this.maximumValue = Math.min(normalizedMaximum, this.max);
+
+      this.#updateHandlesAndTrack();
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (ynotdraw) [reply src/slider.ts:159]: > Should Slider throw, [similar](https://github.com/CrowdStrike/glide-core/blob/main/src/dropdown.ts#L304) to Dropdown, when not multiple and the length of value is longer than 1?

I actually handle this case, and it one I think we should retain to ensure folks can swap between `multiple` and single modes.  I did add a test for it though. We drop off the rest of the array and use the first child. 

https://github.com/CrowdStrike/glide-core/pull/897/commits/51a4b79caeca8a2ff52157b57ed54d57261bc3c8

> Additionally, should Slider throw when multiple and the length of value is greater than 2?

✅ https://github.com/CrowdStrike/glide-core/pull/897/commits/859c3d24799242d3a3cacaede85276ff09f841cf thank you!

File: src/slider.ts (hunk 33 above)
Comment (ynotdraw) [reply src/slider.ts:108]: ✅  https://github.com/CrowdStrike/glide-core/pull/897/commits/6d9e4b43cc1ab581e9edcd6749e36c75fc12b069

File: src/slider.test.events.multiple.ts (hunk 22 above)
Comment (ynotdraw) [reply src/slider.test.events.multiple.ts]: I didn't end up extending `library/mouse.ts`, but ended up using `sendMouse` and our existing `hover` as we talked offline:

✅ https://github.com/CrowdStrike/glide-core/pull/897/commits/cbebc3b763710a11d89cac10165b9522409785f4

Turns out that it appears the `resetMouse()` calls are still needed and the comment is valid 😅

File: src/slider.styles.ts (hunk 14 above)
Comment (ynotdraw) [reply src/slider.styles.ts:121]: Nice. Chatted with Zheng about it, he's open to switching it back to thumb if we feel strongly about it. But I'm fine with "handle", only because we both (design + engineering) are already calling it that at this point.
